torch/utils/data/standard_pipes.ipynb

```diff
@@ -0,0 +1,1006 @@
+     "text": [
+      "[(0, 0), (0, 9), (0, 6), (0, 3)]\n[(1, 4), (1, 7), (1, 1)]\n[(2, 8), (2, 2), (2, 5)]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(10).map(lambda x: (x % 3, x)).shuffle().groupby(lambda x: x[0])\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "By default grouping is applied individually on each item of input mini-batches"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[(1, 4), (1, 1), (1, 10), (1, 7), (1, 13)]\n[(2, 2), (2, 5), (2, 8), (2, 14), (2, 11)]\n[(0, 12), (0, 9), (0, 0), (0, 6), (0, 3)]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(15).map(lambda x: (x % 3, x)).shuffle().groupby(lambda x: x[0])\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "Group key function gets entire mini-batch as input if `batch_level` is True"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[[0, 1, 2], [3, 4, 5], [6, 7, 8]]\n[[9]]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(10).batch(3).groupby(lambda x: len(x), batch_level = True)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "When internal buffer (defined by `buffer_size`) is overfilled, groupby will yield biggest group available"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[(1, 10), (1, 1)]\n[(2, 5), (2, 2), (2, 8)]\n[(0, 6), (0, 3), (0, 0)]\n[(1, 13), (1, 7)]\n[(2, 14), (2, 11)]\n[(0, 9), (0, 12)]\n[(1, 4)]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(15).map(lambda x: (x % 3, x)).shuffle().groupby(lambda x: x[0], buffer_size = 5)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "`groupby` will produce `group_size` sized batches on as fast as possible basis"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 26,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[(1, 1), (1, 4), (1, 7)]\n[(2, 2), (2, 5), (2, 8)]\n[(0, 3), (0, 0), (0, 6)]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(9).map(lambda x: (x % 3, x)).shuffle().groupby(lambda x: x[0], group_size = 3)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "Remaining groups must be at least `guaranteed_group_size` big. "
```

Comment:
curious: how is this guaranteed? Does that mean we only start to emit when the group has `group_size + guaranteed_group_size` elements? (so after we emit `group_size`, there will still be at lest `guaranteed_group_size` left)

Reply:
We accumulate in the buffer all elements until we get at least `guaranteed_group_size` elements. If by the end of stream `guaranteed_group_size` is not reached we raise Exception.